File: Data Structures & Algorithms/matrixBFS/submission-10.py
class Solution:
    def shortestPath(self, grid: List[List[int]]) -> int:
        rows = len(grid)
        cols = len(grid[0])

        if grid[0][0] == 1 or grid[rows - 1][cols - 1] == 1:
            return -1


        steps = 0
        directions = [[1, 0], [-1, 0], [0, 1], [0, -1]]
        q = deque([(0, 0)])
        grid[0][0] = 1

        while q:
            for i in range(len(q)):
                r, c = q.popleft()
                if r == rows - 1 and c == cols - 1:
                    return steps

                for dr, dc in directions:
                    nr = r + dr
                    nc = c + dc
                    if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] != 1:
                        q.append((nr, nc))
                        grid[nr][nc] = 1
            steps += 1

        return -1

File: Data Structures & Algorithms/rotting-fruit/submission-10.ts
class Solution {
    /**
     * @param {number[][]} grid
     * @return {number}
     */
    orangesRotting(grid: number[][]): number {
        const ROWS = grid.length
        const COLS = grid[0].length
        let fresh = 0
        let sources: [number, number][] = []

        for (let r = 0; r < ROWS; r++) {
            for (let c = 0; c < COLS; c++) {
                if (grid[r][c] === 1) {
                    fresh++
                } else if (grid[r][c] === 2) {
                    sources.push([r, c])
                }
            }
        }

        let head = 0
        let time = 0
        const directions = [[1, 0], [0, 1], [-1, 0], [0, -1]]
        
        while (fresh && head < sources.length) {
            let n = sources.length
            for (let i = head; i < n; i++){
                const [r, c] = sources[head++]

                for (const [dr, dc] of directions) {
                    const newRow = r + dr
                    const newCol = c + dc
                    if (newRow >= 0 && newCol >= 0 && newRow < ROWS && newCol < COLS && grid[newRow][newCol] === 1) {
                        grid[newRow][newCol] = 2
                        fresh--
                        sources.push([newRow, newCol])
                    }
                }
            }
            time++
        }
        return fresh === 0 ? time : -1
    }
}
